packages/consensus/fcob/consensusprovider.go

```diff
@@ -0,0 +1,158 @@
+import (
+	"sync"
+
+	"github.com/iotaledger/goshimmer/packages/ledgerstate"
+	"github.com/iotaledger/goshimmer/packages/tangle"
+	"github.com/iotaledger/goshimmer/packages/vote/opinion"
+	voter "github.com/iotaledger/goshimmer/packages/vote/opinion"
+	"github.com/iotaledger/hive.go/events"
+)
+
+type ConsensusProvider struct {
+	Events                 *Events
+	PayloadOpinionProvider *FCoB
+
+	tangle  *tangle.Tangle
+	waiting *opinionWait
+}
+
+func NewConsensusProvider() *ConsensusProvider {
+	return &ConsensusProvider{
+		Events: &Events{
+			PayloadOpinionFormed:   events.NewEvent(payloadOpinionCaller),
+			TimestampOpinionFormed: events.NewEvent(tangle.MessageIDEventHandler),
+		},
+		waiting: &opinionWait{waitMap: make(map[tangle.MessageID]*waitStruct)},
+	}
+}
+
+func (f *ConsensusProvider) Init(tangle *tangle.Tangle) {
+	f.tangle = tangle
+	f.PayloadOpinionProvider = NewFCoB(tangle)
+}
+
+func (f *ConsensusProvider) Setup() {
+	f.tangle.Booker.Events.MessageBooked.Attach(events.NewClosure(f.PayloadOpinionProvider.Evaluate))
+	f.tangle.Booker.Events.MessageBooked.Attach(events.NewClosure(f.EvaluateTimestamp))
+
+	f.PayloadOpinionProvider.Setup(f.Events.PayloadOpinionFormed)
+
+	f.Events.PayloadOpinionFormed.Attach(events.NewClosure(f.onPayloadOpinionFormed))
+}
+
+func (f *ConsensusProvider) EvaluateTimestamp(messageID tangle.MessageID) {
+	f.PayloadOpinionProvider.StoreTimestampOpinion(&TimestampOpinion{
+		MessageID: messageID,
+		Value:     voter.Like,
+		LoK:       Two,
+	})
+
+	f.setEligibility(messageID)
+
+	if f.waiting.done(messageID, timestampOpinion) {
+		f.tangle.OpinionManager.Events.MessageOpinionFormed.Trigger(messageID)
+	}
+}
+
+func (f *ConsensusProvider) PayloadLiked(messageID tangle.MessageID) (liked bool) {
+	return f.PayloadOpinionProvider.Opinion(messageID)
+}
+
+func (f *ConsensusProvider) Shutdown() {
+	f.PayloadOpinionProvider.Shutdown()
+}
+
+func (f *ConsensusProvider) onPayloadOpinionFormed(ev *OpinionFormedEvent) {
+	isTxConfirmed := false
+	// set BranchLiked and BranchFinalized if this payload was a conflict
+	f.tangle.Utils.ComputeIfTransaction(ev.MessageID, func(transactionID ledgerstate.TransactionID) {
+		f.tangle.LedgerState.TransactionMetadata(transactionID).Consume(func(transactionMetadata *ledgerstate.TransactionMetadata) {
+			transactionMetadata.SetFinalized(true)
+		})
+		if f.tangle.LedgerState.TransactionConflicting(transactionID) {
+			f.tangle.LedgerState.BranchDAG.SetBranchLiked(f.tangle.LedgerState.BranchID(transactionID), ev.Opinion)
+			// TODO: move this to approval weight logic
+			f.tangle.LedgerState.BranchDAG.SetBranchFinalized(f.tangle.LedgerState.BranchID(transactionID), true)
+			isTxConfirmed = ev.Opinion
+		}
+	})
+
+	if f.waiting.done(ev.MessageID, payloadOpinion) {
+		f.setEligibility(ev.MessageID)
+		// trigger TransactionOpinionFormed if the message contains a transaction
+		if isTxConfirmed {
+			f.tangle.OpinionManager.Events.TransactionConfirmed.Trigger(ev.MessageID)
+		}
```

Comment:
I have subscribed to this event from the mana plugin to trigger mana update when a transaction is confirmed, but the event doesn't trigger for transactions that are non-conflicting, see line 78 above. @capossele @hans can we set `isTxConfirmed` to `true` here when an opinion is formed on a non-conflicting tx?

Reply:
good catch